File: dotnet/Vuln/Medium/PathTraversalService.cs
namespace SastEval.Vuln.Medium;

/// <summary>
/// CWE-22: Path Traversal - User input used in file path without validation.
/// </summary>
public static class PathTraversalService
{
    private static string ResolvePath(string userInput)
    {
        // VULNERABLE: No validation or canonicalization of user input
        return Path.Combine("/data/uploads", userInput);
    }

    public static IResult Handle(HttpContext context)
    {
        var file = context.Request.Query["file"].ToString();

        if (string.IsNullOrEmpty(file))
        {
            return Results.BadRequest("Missing 'file' query parameter");
        }

        var path = ResolvePath(file);

        try
        {
            // VULNERABLE: Reading file at user-controlled path (e.g., ../../etc/passwd)
            var content = File.ReadAllBytes(path);
            return Results.File(content, "application/octet-stream", Path.GetFileName(path));
        }
        catch (FileNotFoundException)
        {
            return Results.NotFound($"File not found: {file}");
        }
        catch (Exception ex)
        {
            return Results.Problem($"Error reading file: {ex.Message}");
        }
    }
}

File: java/src/main/java/com/sasteval/vuln/medium/PathTraversalService.java
package com.sasteval.vuln.medium;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.OutputStream;

/**
 * CWE-22: Path Traversal
 * VULNERABILITY: User input is used to construct file path without canonicalization.
 */
public class PathTraversalService extends HttpServlet {

    private File resolvePath(String userInput) {
        return new File("/data/uploads/" + userInput);
    }

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {

        String file = request.getParameter("file");

        // VULN: Path traversal via unvalidated user input
        File resolved = resolvePath(file);

        response.setContentType("application/octet-stream");
        response.setHeader("Content-Disposition", "attachment; filename=\"" + resolved.getName() + "\"");

        try (FileInputStream fis = new FileInputStream(resolved);
             OutputStream out = response.getOutputStream()) {

            byte[] buffer = new byte[4096];
            int bytesRead;
            while ((bytesRead = fis.read(buffer)) != -1) {
                out.write(buffer, 0, bytesRead);
            }

        } catch (IOException e) {
            response.setStatus(HttpServletResponse.SC_NOT_FOUND);
            response.getWriter().println("{\"error\": \"File not found\"}");
        }
    }
}
